pkg/provider/kubernetes/alertmanager.go

```diff
@@ -0,0 +1,145 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/provider"
+	"k8c.io/kubermatic/v2/pkg/resources"
+	utilerrors "k8c.io/kubermatic/v2/pkg/util/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// AlertmanagerProvider struct that holds required components in order to manage alertmanager objects.
+type AlertmanagerProvider struct {
+	// createSeedImpersonatedClient is used as a ground for impersonation
+	// whenever a connection to Seed API server is required
+	createSeedImpersonatedClient impersonationClient
+}
+
+// NewAlertmanagerProvider returns an alertmanager provider
+func NewAlertmanagerProvider(createSeedImpersonatedClient impersonationClient) *AlertmanagerProvider {
+	return &AlertmanagerProvider{
+		createSeedImpersonatedClient: createSeedImpersonatedClient,
+	}
+}
+
+func AlertmanagerProviderFactory(mapper meta.RESTMapper, seedKubeconfigGetter provider.SeedKubeconfigGetter) provider.AlertmanagerProviderGetter {
+	return func(seed *kubermaticv1.Seed) (provider.AlertmanagerProvider, error) {
+		cfg, err := seedKubeconfigGetter(seed)
+		if err != nil {
+			return nil, err
+		}
+		defaultImpersonationClientForSeed := NewImpersonationClient(cfg, mapper)
+		return NewAlertmanagerProvider(
+			defaultImpersonationClientForSeed.CreateImpersonatedClient,
+		), nil
+	}
+}
+
+// Get gets an Alertmanager object and Secret which contains the configuration of this Alertmanager.
+func (p *AlertmanagerProvider) Get(cluster *kubermaticv1.Cluster, userInfo *provider.UserInfo) (*kubermaticv1.Alertmanager, *corev1.Secret, error) {
+	impersonationClient, err := createImpersonationClientWrapperFromUserInfo(userInfo, p.createSeedImpersonatedClient)
+	if err != nil {
+		return nil, nil, err
+	}
+	ctx := context.Background()
+	alertmanager := &kubermaticv1.Alertmanager{}
+	if err := impersonationClient.Get(ctx, types.NamespacedName{
+		Name:      resources.AlertmanagerName,
+		Namespace: cluster.Status.NamespaceName,
+	}, alertmanager); err != nil {
+		return nil, nil, err
+	}
+	configSecret := &corev1.Secret{}
+	if err := impersonationClient.Get(ctx, types.NamespacedName{
+		Name:      alertmanager.Spec.ConfigSecret.Name,
+		Namespace: cluster.Status.NamespaceName,
+	}, configSecret); err != nil {
+		return nil, nil, err
+	}
+	return alertmanager, configSecret, nil
+}
+
+// Create only updates an Alertmanager object and corresponding config Secret since Alertmanager and Secret will
+// be created by alertmanager configuration controller.
+func (p *AlertmanagerProvider) Create(expectedAlertmanager *kubermaticv1.Alertmanager, expectedSecret *corev1.Secret, userInfo *provider.UserInfo) (*kubermaticv1.Alertmanager, *corev1.Secret, error) {
```

Comment:
I would say change this method name to update, since it is not really creating resources, it updates them

Reply:
This is actually a good idea, I feel a bit wired when I implemented this `Create` method, WDYT @lsviben @rastislavs @aborilov ?
Also I feel like calling the endpoint `CreateEndpoint` is a bit wired, how about I rename the `CreateEndpoint` to something like `PostEndpoint`?

Reply:
Maybe just make a `UpdateEndpoint` with PUT. You are right its weird to call it Create when it just updates.

Reply:
Thanks @lsviben, `PUT` is also an option. I used `POST` because I would like to keep consistent with cortex alertmanger API: https://cortexmetrics.io/docs/api/#set-alertmanager-configuration
And from user perspective, they are creating/deleting alertmanager configuration, it is just our underneath implementation is doing updating and reseting

Reply:
I see. But if they dont Create anything, they will anyway GET a configuration, the default one, right? 

Same with delete, when they delete it, and then GET, they will get the default config?

Reply:
Yeah you are right, I have replaced `POST` with a `PUT` and also renamed `Create` to `Update` and `Delete` to `Reset` :)